src/main/scala/org/broadinstitute/dsde/workbench/sam/util/ShadowCaster.scala

```diff
@@ -0,0 +1,39 @@
+package org.broadinstitute.dsde.workbench.sam.util
+import cats.effect._
+import com.typesafe.scalalogging.LazyLogging
+
+import scala.concurrent.duration.MILLISECONDS
+import scala.concurrent.{ExecutionContext, Future}
+
+trait ShadowCaster extends LazyLogging {
+  implicit val executionContext: ExecutionContext
+
+  protected def runWithShadow[T](functionName: String, real: IO[T], shadow: IO[T]): IO[T] = {
+
+    val realFuture = measure(real).unsafeToFuture()
+    val shadowFuture = measure(shadow).unsafeToFuture()
+
+    Future.reduceLeft[(T, Long), (T, Long)](List(realFuture, shadowFuture)) { case ((realResult, realTime), (shadowResult, shadowTime)) =>
+      reportResult(functionName, realResult, realTime, shadowResult, shadowTime)
+      (realResult, realTime)
+    }
+
+    IO.fromFuture(IO(realFuture)).map(_._1)
```

Comment:
Some syntactic improvement, you can do
```
    (realFuture, shadowFuture).mapN {case ((realResult, realTime), (shadowResult, shadowTime)) =>
      reportResult(functionName, realResult, realTime, shadowResult, shadowTime)
      (realResult, realTime)
    }
```

I realize yesterday observe won't do exactly what you want.
The following snippet is close to what you have, but not exactly(in this code, reportResult returns IO[Unit]). The difference is real and shadow aren't run in parallel. Shadow runs after real asynchronously, and report stats.

```
    for {
      start <- timer.clock.realTime(MILLISECONDS)
      r <- real
      realEnd <- timer.clock.realTime(MILLISECONDS)
      _ <- shadow.runAsync {
        cb =>
          cb match {
            case Left(error) => IO(logger.error("shadow failed"))
            case Right(t) =>
              for {
                shadowEnd <- timer.clock.realTime(MILLISECONDS)
                _ <- reportResult(functionName, r, realEnd - start, t, shadowEnd - realEnd) //shadow runs after real
              } yield ()
          }
      }.toIO
    } yield r
```
I'll let you know if I think of something better

Reply:
interesting... I had been fixating on running in parallel but I guess we don't need to